docs/howtos/converting-a-component-to-uniffi.md

```diff
@@ -0,0 +1,287 @@
+# Converting an existing Component to use UniFFI
+
+When we started building the components in this repo, exposing Rust code to
+Kotlin and Swift was a manual process and each component had its own
+hand-written FFI layer and foreign-language bindings.
+
+As we've gained more experience with building components in this way, we've
+started to automated bindings generation and capture best practices in a
+tool called [UniFFI](https://mozilla.github.io/uniffi-rs/), which is the
+current recommended approach when [adding a new component from scratch](
+./adding-a-new-component.md).
+
+We expect that existing components will gradually be ported over to use
+UniFFI, and this document is a guide to doing that port.
+
+## First, get familiar with UniFFI
+
+First, make sure you've perused the [UniFFI guide](https://mozilla.github.io/uniffi-rs/)
+to understand the overall architecture of a UniFFI component, and take a look
+at the [guide to adding a new component](./adding-a-new-component.md) to understand
+how such components fit in to this repo. The aim of porting will be to have a component
+that looks like it was added by the process described therein.
+
+## Next, get familiar with the target component
+
+Pre-UniFFI components typically consist of four main parts:
+
+* A Rust crate implementing the core functionality of the component
+* A separate Rust crate that exposes the core functionality over a C-style FFI.
+* An Android package that imports the C-style FFI into idiomatic Kotlin.
+* A Swift module that imports the C-style FFI into idiomatic Swift.
+
+The code for these parts will be laid out something like this:
+
+* `components/<component_name>/`
+    * `Cargo.toml`
+    * `src/`
+        * Rust code for the core functionality of the component goes here.
+    * `ffi/`
+        * `Cargo.toml`
+        * `src/`
+            * Rust code specifically for exposing the C-style FFI goes here.
+    * `android/`
+        * `build.gradle`
+        * `src/`
+          * `main/`
+              * `AndroidManifest.xml`
+              * `java/mozilla/appservices/<component_name>/`
+                  * `LibComponentNameFFI.kt` (low-level bindings to the C-style FFI)
+                  * Higher-level hand-written Kotlin that wraps the FFI.
+    * `ios/`
+        * `<component_name>/`
+          * `Rust<ComponentName>API.h` (low-level bindings to the C-style FFI)
+          * Higher-level hand-written Swift that wraps the FFI.
+
+The goal here is to replace much of the hand-written wrapper layers with autogenerated
+code:
+
+* The `./ffi/` crate will disappear entirely, its work is automated by UniFFI
```

Comment:
Unless I misunderstand, if you are only doing a partial conversion, this might remain, right? If that's true, I think it might be worth setting the scene for that above - there are lots of places below which assume "will be removed", so I don't think it's worth trying to qualify that everwhere, but calling it out as an intro to this section might make sense.

Reply:
That's a good point, I hadn't really considered partial conversions here...I guess that would be for the case if there's something that UniFFI doesn't support yet, and you need to write your own `pub extern "C"` functions?

I think I would still argue that those functions should live in the main crate and be part of `lib.rs` and your top-level public API, rather than squirreled away in a separate `./ffi/` crate. But I could add some words to this effect if you think it would be helpful?

(In fact I don't really recall why we settled on this convention of having a separate `ffi` crate at all to be honest...)

Reply:
Yes, that would be perfect - just a small note saying "things which are unable to be converted should be moved into lib.rs" would be perfect and about all the detail needed here.